README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# ppmf <a href='https://www.christophertkenny.com/ppmf/'><img src='man/figures/logo.png' align="right" height="129" /></a>

<!-- badges: start -->

[![R-CMD-check](https://github.com/christopherkenny/ppmf/workflows/R-CMD-check/badge.svg)](https://github.com/christopherkenny/ppmf/actions)
<!-- badges: end -->

The goal of ppmf is to convert Census Privacy Protected Microdata Files
into somewhat wider data aggregated to a geographic level.

## Installation

You can install the development version from
[GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("christopherkenny/ppmf")
```

## Basics

Load the package:

``` r
library(ppmf)
```

Download and read data with:

``` r
path <- download_ppmf(dsn = 'filename.csv', dir = 'some/directory', version = '19')
al <- read_ppmf(state = 'AL', path = path)
```

Version ‘19’ reflects the 19.61 finalized parameters used again for the
2020 Census.

For future use, I recommend storing the path to the data for future
sessions using:

``` r
add_pmmf19_path(path)
```

Then the path can be recovered with:

``` r
path19 <- Sys.getenv('ppmf19')
```

Once you’ve read in what you want, you can aggregate it to the right
level:

``` r
al <- al %>% add_geoid()
blocks <- agg(al)
```

And aggregated data can use the GEOID to merge with shapefiles:

``` r
library(dplyr) # to clean up the data

shp <- tigris::blocks('AL', year = 2010) %>% 
  select(GEOID10, geometry) %>% rename(GEOID = GEOID10)
shp <- shp %>% left_join(blocks, by = 'GEOID')

# always clean shp!
shp[is.na(shp)] <- 0
```

------------------------------------------------------------------------

For users with the newest package version, there is an added dependency
on [`censable`](https://www.christophertkenny.com/censable/), which
allows for an easier workflow. If you’ve used the `add_pmmf*_path()`
workflow suggested, you don’t even need to supply the paths!

This will not just read the `ppmf` data, it will merge it with 2010
Census populations (by major race/ethnicity grouping) and add the
corresponding geometries.

``` r
al <- read_merge_ppmf('AL', level = 'block', versions = '19')
```


NEWS.md
# ppmf 0.1.1
* Add default support for PLB 19.61 to `read_merge_ppmf()`.

# ppmf 0.1.0
* Add support for PLB 19.61

# ppmf 0.0.6
* Adds `censable` dependency and function `read_merge_ppmf` to create cleaner merging workflows.
* Removes references to the string `block_group`, moving towards the more consistent `block group`.

# ppmf 0.0.5
* Fix issue where `read_ppmf()` can't read in files if there's an object named `states` in the global environment.

# ppmf 0.0.4
* Update information for public use

# ppmf 0.0.3
* Uses GitHub release links for compressed data
* Add get_ppmf_links() to get the links for data
* Basic pkgdown site
* Function to add ppmf paths to environment

# ppmf 0.0.2
* Added a `NEWS.md` file to track changes to the package.
* Added function breakdown_geoid() to convert GEOIDs into component parts
* fix typo in ppmf_ex names
* Runs tools::resaveRdaFiles('data/') to compress Rda for faster install


R/data.R

#' Example PPMF Data
#'
#' @name ppmf_ex
#'
#' @description
#' Includes Perry County, Alabama PPMF data from the April 28, 2021 PPMF data release.
#' This is a subset taken from the 12-2 P data.
#'
#' As each observation is a person, this does not cover every block in the county and
#' due to DAS, not every block with population appears in this data.
#'
#' @return tibble with sample ppmf data
#'
#' @usage
#' data('ppmf_ex')
#'
#' @examples
#' data('ppmf_ex')
#' 
#' @concept data
NULL

#' State Rows
#'
#' @name states
#'
#' @description
#' This data includes the 52 geographies (50 states plus D.C. and P.R.). Within the
#' 2010 PPMF, skip and n_max indicate the relevant rows for a geography.
#'
#' @return tibble with sample ppmf data
#'
#' @usage
#' data('states')
#'
#' @examples
#' data('states')
#' 
#' @concept data
NULL

#' Race Classifications
#' @name races
#'
#' @description
#' This data includes the basic race classifications used for redistricting to
#' get to an easier to work with set of values. This does not include `hisp` grouping
#' which is controlled separately by race within the census
#'
#' @return tibble with three columns
#'  - code: the two digit code used to code races
#'  - desc: the description of the races
#'  - group: the summary group used
#'
#' @usage
#' data('races')
#'
#' @examples
#' data('races')
#'
#'
#' @concept data
#' @md
NULL


R/race.R
#' Replace Race Categories
#'
#' Replaces the Census's numeric categories for race with less specific racial
#' classifications, typically useful for redistricting purposes.
#'
#' @param ppmf tibble of ppmf data
#' @param race Column in ppmf containing race codes
#'
#' @return tibble with race column replaced by simpler racial classifications
#' @export
#'
#' @importFrom dplyr .data rename select left_join
#' @importFrom rlang as_name
#' @concept basic
#' @examples
#' data(ppmf_ex)
#' ppmf_ex %>% replace_race()
replace_race <- function(ppmf, race = CENRACE){
  races <- get('races') %>% select(-.data$desc)

  #races <- races %>% rename(race = .data$code)
  races[[as_name(enquo(race))]] <- races$code

  if(as_name(enquo(race)) != 'code'){
    races <- races %>% select(-.data$code)
  }

  races <- races %>% rename(group_race_name = .data$group)

  ppmf <- ppmf %>% left_join(races, by = as_name(enquo(race)))

  ppmf$group_race_name[is.na(ppmf$group_race_name)] <- 'hisp'

  ppmf[[as_name(enquo(race))]] <- ppmf$group_race_name
  ppmf[['group_race_name']] <- NULL


  return(ppmf)
}


#' Overwrite Races with Hispanic
#'
#' @param ppmf tibble of ppmf data
#' @param race Column in ppmf containing race codes
#' @param hisp Column in ppmf containing 1 for Not Hispanic and 2 for Hispanic
#'
#' @return tibble with race column entries replaced if the individual is Hispanic
#' @export
#' @concept basic
#' @examples
#' data(ppmf_ex)
#' ppmf_ex %>% replace_race() %>% overwrite_hisp_race()
overwrite_hisp_race <- function(ppmf, race = CENRACE, hisp = CENHISP){

  ppmf[[as_name(enquo(race))]][ppmf[[as_name(enquo(hisp))]] == 2 ] <- 'hisp'

  return(ppmf)

}

utils::globalVariables(c('CENRACE', 'CENHISP'))


R/geoid.R

#' Add Standard GEOID to PPMF Data
#'
#' Adds the GEOID identifier common to spatial census data sets, such as those loaded
#' by tigris. This allows for easier merging or aggregation by a single variable.
#'
#' @param ppmf tibble of ppmf data
#' @param state Column in ppmf with state (fips) ID. Default is \code{TABBLKST}.
#' @param county Column in ppmf with county (fips) ID. Default is \code{TABBLKCOU}.
#' @param tract Column in ppmf with tract ID. Default is \code{TABBLKTRACT}.
#' @param block_group Column in ppmf with block group ID. Default is \code{TABBLKGRP}
#' @param block Column in ppmf with block ID. Default is \code{TABBLK}.
#' @param level Geographic level to write the GEOID for. Options are block (default),
#' block_group, tract, and county.
#'
#' @return input data ppmf with added column GEOID
#' @export
#'
#' @importFrom rlang enquo eval_tidy
#' @importFrom dplyr mutate
#' @concept geoid
#' @examples
#' data(ppmf_ex)
#' ppmf_ex <- ppmf_ex %>% add_geoid()
#'
add_geoid <- function(ppmf, state = TABBLKST, county = TABBLKCOU,
                      tract = TABTRACT, block_group = TABBLKGRP,
                      block = TABBLK, level = 'block'){

  if(missing(ppmf)){
    stop('ppmf argument missing in add_geoid.')
  }

  match.arg(level, choices = c('block', 'block group', 'tract', 'county'))

  if( level == 'block' ){
    block_group <- NULL
  } else if ( level == 'block group' ) {
    block <- NULL
  } else if ( level == 'tract') {
    block <- NULL
    block_group <- NULL
  } else {
    block <- NULL
    block_group <- NULL
    tract <- NULL
  }

  ppmf %>% mutate(GEOID = paste0({{state}}, {{county}}, {{tract}}, {{block_group}},
                                 {{block}}))

}

#' Breakdown GEOID into Components
#'
#' @param ppmf tibble of ppmf data
#' @param GEOID Column in ppmf with GEOID. Default is \code{GEOID}.
#'
#' @return tibble. ppmf with columns added for state, county, tract, block group, and/or block
#' @export
#'
#' @importFrom stringr str_sub str_length
#' @concept geoid
#' @examples
#' data(ppmf_ex)
#' ppmf_ex <- ppmf_ex %>% add_geoid()
#' ppmf_ex <- ppmf_ex %>% breakdown_geoid()
breakdown_geoid <- function(ppmf, GEOID = GEOID){
  if(missing(ppmf)){
    stop('ppmf argument missing in add_geoid.')
  }

  geoid_col <- eval_tidy(enquo(GEOID), ppmf)

  if(is.null(geoid_col[1])){
    stop('`GEOID` is not a column in ppmf.')
  }

  if(str_length(geoid_col[1]) < 2){
    stop('GEOID does not have a recognizable pattern.')
  } else {
    len <- str_length(geoid_col[1])
  }

  ppmf <- ppmf %>% mutate(state = str_sub({{GEOID}}, 1, 2))

  if(len >= 5){
    ppmf <- ppmf %>% mutate(county = str_sub({{GEOID}}, 3, 5))
  }

  if(len >= 11){
    ppmf <- ppmf %>% mutate(tract = str_sub({{GEOID}}, 6, 11))
  }

  if(len >= 12){
    ppmf <- ppmf %>% mutate(block_group = str_sub({{GEOID}},12,12))
  }
  if(len >= 15){
    ppmf <- ppmf %>% mutate(block = str_sub({{GEOID}}, 12, 15))
  }

  ppmf
}

utils::globalVariables(c('TABBLKST', 'TABBLKCOU', 'TABTRACT',
                         'TABBLKGRP', 'TABBLK', 'GEOID'))


R/add_path.R
#' Add ppmf12 path to Renviron
#'
#' @param path path where ppmf12 data is stored
#' @param overwrite Defaults to FALSE. Should existing ppmf12 in Renviron be overwritten?
#' @param install Defaults to FALSE. Should ppmf12 be added to '~/.Renviron' file?
#'
#' @return path, invisibly
#' @export
#' @concept save
#' @examples
#' \dontrun{
#' tp <- tempfile(fileext = '.csv')
#' add_ppmf12_path(tp)
#' path12 <- Sys.getenv('path12')
#' }
#'
add_ppmf12_path <- function(path, overwrite = FALSE, install = FALSE) {
  if (missing(path)) {
    stop('Input `path` cannot be missing.')
  }

  if (install) {
    r_env <- file.path(Sys.getenv('HOME'), '.Renviron')

    if (!file.exists(r_env)) {
      file.create(r_env)
    }

    lines <- readLines(r_env)
    newline <- paste0("ppmf12='", path.expand(path), "'")

    exists <- stringr::str_detect(lines, 'ppmf12=')

    if (any(exists)) {
      if (sum(exists) > 1) {
        stop('Multiple entries in .Renviron have name matching input `name`.\nEdit manually with `usethis::edit_r_environ()`.')
      }

      if (overwrite) {
        lines[exists] <- newline
        writeLines(lines, r_env)
        message('Run `readRenviron("~/.Renviron")` to update your active environment.')
      } else {
        message('ppmf12 already exists in .Renviron. \nEdit manually with `usethis::edit_r_environ() or set `overwrite = TRUE`.')
      }
    } else {
      lines[length(lines) + 1] <- newline
      writeLines(lines, r_env)
      message('Run `readRenviron("~/.Renviron")` to update your active environment.')
    }
  } else {
    Sys.setenv(ppmf12 = path)
  }

  invisible(path)
}


#' Add ppmf4 path to Renviron
#'
#' @param path path where ppmf4 data is stored
#' @param overwrite Defaults to FALSE. Should existing ppmf4 in Renviron be overwritten?
#' @param install Defaults to FALSE. Should ppmf4 be added to '~/.Renviron' file?
#'
#' @return path, invisibly
#' @export
#'
#' @concept save
#' @examples
#' \dontrun{
#' tp <- tempfile(fileext = '.csv')
#' add_ppmf4_path(tp)
#' path4 <- Sys.getenv('path4')
#' }
#'
add_ppmf4_path <- function(path, overwrite = FALSE, install = FALSE) {
  if (missing(path)) {
    stop('Input `path` cannot be missing.')
  }

  if (install) {
    r_env <- file.path(Sys.getenv('HOME'), '.Renviron')

    if (!file.exists(r_env)) {
      file.create(r_env)
    }

    lines <- readLines(r_env)
    newline <- paste0("ppmf4='", path.expand(path), "'")

    exists <- stringr::str_detect(lines, 'ppmf4=')

    if (any(exists)) {
      if (sum(exists) > 1) {
        stop('Multiple entries in .Renviron have name matching input `name`.\nEdit manually with `usethis::edit_r_environ()`.')
      }

      if (overwrite) {
        lines[exists] <- newline
        writeLines(lines, r_env)
        message('Run `readRenviron("~/.Renviron")` to update your active environment.')
      } else {
        message('ppmf4 already exists in .Renviron. \nEdit manually with `usethis::edit_r_environ() or set `overwrite = TRUE`.')
      }
    } else {
      lines[length(lines) + 1] <- newline
      writeLines(lines, r_env)
      message('Run `readRenviron("~/.Renviron")` to update your active environment.')
    }
  } else {
    Sys.setenv(ppmf4 = path)
  }

  invisible(path)
}


#' Add ppmf19 path to Renviron
#'
#' @param path path where ppmf19 data is stored
#' @param overwrite Defaults to FALSE. Should existing ppmf19 in Renviron be overwritten?
#' @param install Defaults to FALSE. Should ppmf19 be added to '~/.Renviron' file?
#'
#' @return path, invisibly
#' @export
#' @concept save
#' @examples
#' \dontrun{
#' tp <- tempfile(fileext = '.csv')
#' add_ppmf19_path(tp)
#' path19 <- Sys.getenv('path19')
#' }
#'
add_ppmf19_path <- function(path, overwrite = FALSE, install = FALSE) {
  if (missing(path)) {
    stop('Input `path` cannot be missing.')
  }

  if (install) {
    r_env <- file.path(Sys.getenv('HOME'), '.Renviron')

    if (!file.exists(r_env)) {
      file.create(r_env)
    }

    lines <- readLines(r_env)
    newline <- paste0("ppmf19='", path.expand(path), "'")

    exists <- stringr::str_detect(lines, 'ppmf19=')

    if (any(exists)) {
      if (sum(exists) > 1) {
        stop('Multiple entries in .Renviron have name matching input `name`.\nEdit manually with `usethis::edit_r_environ()`.')
      }

      if (overwrite) {
        lines[exists] <- newline
        writeLines(lines, r_env)
        message('Run `readRenviron("~/.Renviron")` to update your active environment.')
      } else {
        message('ppmf19 already exists in .Renviron. \nEdit manually with `usethis::edit_r_environ() or set `overwrite = TRUE`.')
      }
    } else {
      lines[length(lines) + 1] <- newline
      writeLines(lines, r_env)
      message('Run `readRenviron("~/.Renviron")` to update your active environment.')
    }
  } else {
    Sys.setenv(ppmf19 = path)
  }

  invisible(path)
}


R/aggregate.R
#' Aggregate PPMF Data
#'
#' @param ppmf tibble of ppmf data
#' @param group Column in ppmf to group by, typically GEOID
#' @param age Column in ppmf containing 1 for not voting age and 2 for voting age
#' @param race Column in ppmf containing race codes
#' @param hisp Column in ppmf containing 1 for Not Hispanic and 2 for Hispanic
#'
#' @return tibble of ppmf data aggregated by group with race classified with columns:
#' * `group`: named by entry group
#' * `pop`: total population
#' * `pop_hisp`: total population - Hispanic or Latino (of any race)
#' * `pop_white`: total population - White alone, not Hispanic or Latino
#' * `pop_black`: total population - Black or African American alone, not Hispanic or Latino
#' * `pop_aian`: total population - American Indian and Alaska Native alone, not Hispanic or Latino
#' * `pop_asian`: total population - Asian alone, not Hispanic or Latino
#' * `pop_nhpi`: total population - Native Hawaiian and Other Pacific Islander alone, not Hispanic or Latino
#' * `pop_other`: total population - Some Other Race alone, not Hispanic or Latino
#' * `pop_two`: total population - Population of two or more races, not Hispanic or Latino
#' * `vap`: voting age population
#' * `vap_hisp`: voting age population - Hispanic or Latino (of any race)
#' * `vap_white`: voting age population - White alone, not Hispanic or Latino
#' * `vap_black`: voting age population - Black or African American alone, not Hispanic or Latino
#' * `vap_aian`: voting age population - American Indian and Alaska Native alone, not Hispanic or Latino
#' * `vap_asian`: voting age population - Asian alone, not Hispanic or Latino
#' * `vap_nhpi`: voting age population - Native Hawaiian and Other Pacific Islander alone, not Hispanic or Latino
#' * `vap_other`: voting age population - Some Other Race alone, not Hispanic or Latino
#' * `vap_two`: voting age population - Population of two or more races, not Hispanic or Latino
#'
#'
#' @export
#'
#' @importFrom dplyr .data select full_join count group_by filter rowwise starts_with ungroup
#' @importFrom tidyr pivot_wider
#' @importFrom tibble add_column
#' 
#' @concept basic
#' @md
#'
#' @examples
#' data(ppmf_ex)
#' ppmf_ex <- ppmf_ex %>% add_geoid()
#' blocks <- agg(ppmf_ex)
#'
agg <- function(ppmf, group = GEOID, age = VOTING_AGE, race = CENRACE, hisp = CENHISP){
  ppmf <- ppmf %>% select({{group}}, {{age}}, {{race}}, {{hisp}})

  ppmf <-  ppmf %>%
    replace_race(race = !!enquo(race)) %>%
    overwrite_hisp_race(race = !!enquo(race), hisp = !!enquo(hisp))

  ppmf <- ppmf %>% select(-{{hisp}})



  ppmf_vap <- ppmf %>% filter({{age}} == 2) %>% select(-{{age}}) %>%
    group_by({{group}}) %>% count({{race}}) %>%
    pivot_wider(id_cols = {{group}}, names_from = {{race}},
                values_from = .data$n, values_fill = 0, names_prefix = 'vap_')


  ppmf_nvap <- ppmf %>% filter({{age}} == 1) %>% select(-{{age}}) %>%
    group_by({{group}}) %>% count({{race}}) %>%
    pivot_wider(id_cols = {{group}}, names_from = {{race}},
                values_from = .data$n, values_fill = 0, names_prefix = 'nvap_')

  # ensure all needed columns present
  exp_cols <- c('black', 'white', 'hisp', 'two', 'other', 'aian', 'nhpi', 'asian')
  exp_vap <- paste0('vap_', exp_cols)
  exp_nvap <- paste0('nvap_', exp_cols)

  add_vap <- rep(0, length(exp_vap))
  names(add_vap) <- exp_vap
  ppmf_vap <- ppmf_vap %>% add_column(!!!add_vap[setdiff(names(add_vap),
                                                         names(ppmf_vap))])
  add_nvap <- rep(0, length(exp_nvap))
  names(add_nvap) <- exp_nvap
  ppmf_nvap <- ppmf_nvap %>% add_column(!!!add_nvap[setdiff(names(add_nvap),
                                                           names(ppmf_nvap))])

  # join

  ret <- ppmf_vap %>% full_join(ppmf_nvap, by = 'GEOID')

  # get the correct objects
  ret[is.na(ret)] <- 0
  ret <- ret %>% ungroup()
  ret <- ret %>% mutate(nvap = rowSums(select(., starts_with('nvap')), na.rm = TRUE),
                                      vap = rowSums(select(., starts_with('vap')), na.rm = TRUE))

  ret <- ret %>% mutate(pop = .data$nvap + .data$vap,
                        pop_white = .data$vap_white + .data$nvap_white,
                        pop_hisp = .data$vap_hisp + .data$nvap_hisp,
                        pop_black = .data$vap_black + .data$nvap_black,
                        pop_aian = .data$vap_aian + .data$nvap_aian,
                        pop_asian = .data$vap_asian + .data$nvap_asian,
                        pop_nhpi = .data$vap_nhpi + .data$nvap_nhpi,
                        pop_other = .data$vap_other + .data$nvap_other,
                        pop_two = .data$vap_two + .data$nvap_two)
  # reorder and return
  ret %>% select(.data$GEOID,
                 .data$pop, .data$pop_white, .data$pop_hisp,
                 .data$pop_black, .data$pop_aian, .data$pop_asian,
                 .data$pop_nhpi, .data$pop_other, .data$pop_two,
                 .data$vap, .data$vap_white, .data$vap_hisp,
                 .data$vap_black, .data$vap_aian, .data$vap_asian,
                 .data$vap_nhpi, .data$vap_other, .data$vap_two)

}

utils::globalVariables(c('GEOID','VOTING_AGE', '.'))


R/read_merge.R
#' Read PPMF data and Merge with Census 2010 Data
#'
#' @param state state abbreviation
#' @param level geography level. One of 'block', 'block group', 'tract', 'county'
#' @param versions character vector of ppmf versions. Currently '19', '12', and/or '4'
#' @param prefixes prefixes to give pop and vap columns in output. Default is `paste0('v', versions, '_')`
#' @param paths paths to PPMF data. Default is `Sys.getenv(paste0('ppmf', versions))`
#'
#' @return sf tibble of PPMF merged with Census 2010 data
#' @export
#'
#' @concept getdata
#' @md
#' @examples
#' \dontrun{
#' # Requires Census Bureau API
#' de_bg <- read_merge_ppmf('DE', 'block group')
#' }
read_merge_ppmf <- function(state, level, versions = c('19'),
                            prefixes = paste0('v', versions, '_'),
                            paths = Sys.getenv(paste0('ppmf', versions))) {

  # check inputs ----
  state <- censable::match_abb(state)
  stopifnot(length(state) == 1)
  match.arg(level, choices = c('block', 'block group', 'tract', 'county'))
  match.arg(versions, choices = c('19', '12', '4'), several.ok = TRUE)

  if (length(versions) != length(prefixes)) {
    stop('`versions` and `prefixes` must have equal lengths.')
  }

  lapply(seq_along(paths), function(x){
    if (!file.exists(paths[x])) {
      stop(paste0('File ', paths[x], 'does not exist. Did you download it with `download_ppmf`?'))
    }
  })

  # read census data ----
  cen <- censable::build_dec(geography = level, state = state, year = 2010)

  # read ppmf data ----
  ppmfs <- lapply(seq_along(versions), function(x) {
    read_ppmf(state = state, path = paths[x]) %>%
    add_geoid(level = level) %>%
    agg() %>%
    dplyr::rename_with(.fn = function(i){paste0(prefixes[x], i)},
                       dplyr::starts_with(c('pop', 'vap')))
    })

  for (i in seq_along(ppmfs)) {
    cen <- cen %>%
      left_join(ppmfs[[i]], by = 'GEOID')
  }

  cen[is.na(cen)] <- 0

  cen
}


R/get_data.R
#' Read in PPMF Data
#'
#' This reads in PPMF data from a file. Use `download_ppmf()` if you do
#' not have a local copy of the ppmf data.
#'
#' @param state two letter state (+ DC + PR) abbreviation or two digit state fips code
#' @param path where the data is saved to
#'
#' @return tibble of ppmf data
#' @export
#'
#' @importFrom readr read_csv
#' @importFrom stringr str_pad
#'
#' @concept getdata
#'
#' @examples
#' \dontrun{
#' # Takes a few minutes and requires read access to files
#' temp <- tempdir()
#' path <- download_ppmf('ppmf_12.csv', dir = temp)
#' # If you already have it downloaded, point to it with path:
#' ppmf <- read_ppmf('AL', path)
#' }
read_ppmf <- function(state, path){
  states <- ppmf::states

  if(is.numeric(state)){
    state <- str_pad(state, 2, 'left', '0')
  }
  which_state <- which(state == states$state)
  if(length(which_state) == 0){
    which_state <- which(state == states$state_code)
  }
  if(length(which_state) == 0){
    stop('state does not match a state abbreviation or state fips.\n
         Check `data(states)` columns state or state_code.')
  }


  if(!file.exists(path)){
    stop('`path` does not point to a file.')
  }


  cols_nom <- c('VINTAGE', 'TABBLKST', 'TABBLKCOU',
                'TABTRACT', 'TABBLKGRP', 'TABBLK',
                'RTYPE', 'GQTYPE_PL', 'VOTING_AGE',
                'CENHISP', 'CENRACE')

  read_csv(path,
           skip = states$skip[which_state],
           n_max = states$n_max[which_state],
           col_types = 'ccccccccccc',
           col_names = cols_nom)



}

#' Download PPMF Files
#'
#'  Downloads zipped ppmf files from GitHub.
#'
#' @param dsn (data save name) string to unzip the data to
#' @param dir the folder or directory to save the file in
#' @param version string in '19', '12' or '4' signifying the 19.61, 12.2 or 4.5 versions respectively
#' @param overwrite If a file is found at path/dsn, should it be overwritten?
#' Defaults to FALSE.
#'
#' @return a string path to where the file was downloaded to
#' @export
#'
#' @importFrom utils download.file
#' @importFrom zip unzip
#'
#' @concept getdata
#'
#' @examples
#' \dontrun{
#' # Takes a few minutes and requires read access to files
#' temp <- tempdir()
#' path <- download_ppmf(dsn = 'ppmf_12', dir = temp)
#' }
download_ppmf <- function(dsn, dir = '', version = '19', overwrite = FALSE){
  match.arg(version, choices = c('19','12', '4'))

  if(!missing(dsn)){
    if(stringr::str_sub(dsn, -4, -1) == '.csv'){
      if(dir == ''){
        path <- dsn
      } else {
        path <- paste0(dir, '/', dsn, '.csv')
      }
    } else {
      if(dir == ''){
        path <- paste0(dsn, '.csv')
      } else {
        path <- paste0(dir, '/', dsn, '.csv')
      }
    }
  } else { # dsn is missing
    dsn <- paste0('ppmf_', version, '.csv')
    if(dir == ''){
      path <- dsn
    } else {
      path <- paste0(dir, '/', dsn, '.csv')
    }
  }

  if(file.exists(path) & !overwrite){
    cat('File already exists at the given path. To overwrite, set `overwrite = TRUE`.')
    return(path)
  }

  # If there is no file or we need to overwrite it, download zip:
  temp <- tempfile(fileext = '.zip')

  zip_path <- 'https://github.com/christopherkenny/ppmf_data/releases/download/08122021/ppmf_19.zip'
  if(version == '12') {
    zip_path <- 'https://github.com/christopherkenny/ppmf_data/releases/download/04282021/ppmf_12.zip'
  } else if (version == '4') {
    zip_path <- 'https://github.com/christopherkenny/ppmf_data/releases/download/04282021/ppmf_4.zip'
  }


  download.file(zip_path, temp)

  # then create the desired file:
  cat('File downloaded. Unzipping, please wait.\n')
  zip::unzip(zipfile = temp, exdir = dir)

  # rename it
  down_name <- 'ppmf_20210608_P.csv'
  if (version == '12') {
    down_name <- 'ppmf_20210428_eps12-2_P.csv'

  } else if (version == '4') {
    down_name <- 'ppmf_20210428_eps4-5_P.csv'
  }


  file.rename(from = stringr::str_glue('{dir}/{down_name}'), to = path)

  # and return where we put it:
  return(path)

}


#' Get PPMF File Links
#'
#' Returns the urls for the data. This will be expanded to link to prior or
#' any new releases.
#'
#'
#' @param version string in '19', '12' or '4' signifying the 19.61, 12.2, or 4.5 versions respectively
#' @param release string. Ignored. Options are '06.08.2021' and '04.28.2021'.
#' @param compressed boolean. Return a compressed version (TRUE). FALSE gives the
#' Census Bureau link to the uncompressed data.
#' @return a string with url
#' @export
#'
#' @concept getdata
#'
#' @examples
#' # 04.28.2021 version 12.2
#' get_ppmf_links()
#' # 04.28.2021 version 4.5
#' get_ppmf_links(version = '4')
get_ppmf_links <- function(version = '19', release = '06.08.2021', compressed = TRUE){
  match.arg(version, choices = c('19', '12', '4'))

  # if(release == '04.28.2021'){
  if(compressed){
    if (version == '12') {
      'https://github.com/christopherkenny/ppmf_data/releases/download/04282021/ppmf_12.zip'
    } else if (version == '4') {
      'https://github.com/christopherkenny/ppmf_data/releases/download/04282021/ppmf_4.zip'
    } else {
      'https://github.com/christopherkenny/ppmf_data/releases/download/08122021/ppmf_19.zip'
    }
  } else {
    if (version == '12') {
      'https://www2.census.gov/programs-surveys/decennial/2020/program-management/data-product-planning/2010-demonstration-data-products/ppmf20210428/ppmf_20210428_eps12-2_P.csv'
    } else if (version == '4') {
      'https://www2.census.gov/programs-surveys/decennial/2020/program-management/data-product-planning/2010-demonstration-data-products/ppmf20210428/ppmf_20210428_eps4-5_P.csv'
    } else {
      'https://www2.census.gov/programs-surveys/decennial/2020/program-management/data-product-planning/2010-demonstration-data-products/ppmf20210608/ppmf_20210608_P.csv'
    }
  }
  #}

}
